frame/support/src/storage/bounded_vec.rs

```diff
@@ -0,0 +1,443 @@
+	///
+	/// This might fail if bounds are not respected.
+	fn try_append<LikeT: EncodeLike<T>>(item: LikeT) -> Result<(), ()>;
+}
+
+/// Storage map that is *maybe* capable of [`StorageAppend`].
+pub trait TryAppendMap<K: FullCodec, T: BoundedVecValue, S: Get<u32>> {
+	/// Try and append the `item` into the storage map at the given `key`.
+	///
+	/// This might fail if bounds are not respected.
+	fn try_append<LikeK: EncodeLike<K> + Clone, LikeT: EncodeLike<T>>(
+		key: LikeK,
+		item: LikeT,
+	) -> Result<(), ()>;
+}
+
+/// Storage double map that is *maybe* capable of [`StorageAppend`].
+pub trait TryAppendDoubleMap<K1: FullCodec, K2: FullCodec, T: BoundedVecValue, S: Get<u32>> {
+	/// Try and append the `item` into the storage double map at the given `key`.
+	///
+	/// This might fail if bounds are not respected.
+	fn try_append<
+		LikeK1: EncodeLike<K1> + Clone,
+		LikeK2: EncodeLike<K2> + Clone,
+		LikeT: EncodeLike<T>,
+	>(
+		key1: LikeK1,
+		key2: LikeK2,
+		item: LikeT,
+	) -> Result<(), ()>;
+}
+
+impl<T: BoundedVecValue, S: Get<u32>, StorageValueT: generator::StorageValue<BoundedVec<T, S>>>
+	TryAppendValue<T, S> for StorageValueT
+{
+	fn try_append<LikeT: EncodeLike<T>>(item: LikeT) -> Result<(), ()> {
+		let bound = BoundedVec::<T, S>::bound();
+		let current = Self::decode_len().unwrap_or_default();
+		if current < bound {
+			// NOTE: we cannot reuse the implementation for `Vec<T>` here because we never want to
+			// mark `BoundedVec<T, S>` as `StorageAppend`.
+			let key = Self::storage_value_final_key();
+			sp_io::storage::append(&key, item.encode());
+			Ok(())
+		} else {
+			Err(())
+		}
+	}
+}
+
+impl<
+		K: FullCodec,
+		T: BoundedVecValue,
+		S: Get<u32>,
+		StorageMapT: generator::StorageMap<K, BoundedVec<T, S>>,
+	> TryAppendMap<K, T, S> for StorageMapT
+{
+	fn try_append<LikeK: EncodeLike<K> + Clone, LikeT: EncodeLike<T>>(
+		key: LikeK,
+		item: LikeT,
+	) -> Result<(), ()> {
+		let bound = BoundedVec::<T, S>::bound();
+		let current = Self::decode_len(key.clone()).unwrap_or_default();
+		if current < bound {
+			let key = Self::storage_map_final_key(key);
+			sp_io::storage::append(&key, item.encode());
+			Ok(())
+		} else {
+			Err(())
+		}
+	}
+}
+
+impl<
+		K1: FullCodec,
+		K2: FullCodec,
+		T: BoundedVecValue,
+		S: Get<u32>,
+		StorageDoubleMapT: generator::StorageDoubleMap<K1, K2, BoundedVec<T, S>>,
+	> TryAppendDoubleMap<K1, K2, T, S> for StorageDoubleMapT
+{
+	fn try_append<
+		LikeK1: EncodeLike<K1> + Clone,
+		LikeK2: EncodeLike<K2> + Clone,
+		LikeT: EncodeLike<T>,
+	>(
+		key1: LikeK1,
+		key2: LikeK2,
+		item: LikeT,
+	) -> Result<(), ()> {
+		let bound = BoundedVec::<T, S>::bound();
+		let current = Self::decode_len(key1.clone(), key2.clone()).unwrap_or_default();
+		if current < bound {
+			let double_map_key = Self::storage_double_map_final_key(key1, key2);
+			sp_io::storage::append(&double_map_key, item.encode());
+			Ok(())
+		} else {
+			Err(())
+		}
+	}
+}
+
+#[cfg(test)]
+pub mod test {
```

Comment:
maybe add some tests for the bounded vec itself?

Reply:
Yeah, I should add a test for anything that is not coerced via the deref.

Reply:
added a bunch, but not for the functions like `retain` that only relay the call to `self.0`. Then I feel like I am writing unit tests for rust's `Vec<_>`, and thats not our responsibility + it is a big rabbit hole :D